src/client/common/process/pythonDaemon.ts

```
@@ -27,22 +28,36 @@ import {
 
 type ErrorResponse = { error?: string };
 
+export class ConnectionClosedError extends Error {
+    constructor(public readonly message: string){
+        super();
+    }
+}
 export class PythonDaemonExecutionService implements IPythonDaemonExecutionService {
-    private connectionClosedMessage?: string;
+    private connectionClosedMessage: string = '';
     private outputObservale = new Subject<Output<string>>();
+    // tslint:disable-next-line: no-any
+    private readonly connectionClosedDeferred: Deferred<any>;
     private disposables: IDisposable[] = [];
+    public get isAlive(): boolean {
+        return this.connectionClosedMessage === '';
+    }
     constructor(
         protected readonly pythonExecutionService: IPythonExecutionService,
         protected readonly pythonPath: string,
-        protected readonly daemonProc: ChildProcess,
-        protected readonly connection: MessageConnection
+        public readonly proc: ChildProcess,
+        public readonly connection: MessageConnection
     ) {
+        // tslint:disable-next-line: no-any
+        this.connectionClosedDeferred = createDeferred<any>();
+        // We don't want any unhandled exceptions from this promise.
+        this.connectionClosedDeferred.promise.ignoreErrors();
```

>ignoreErrors [](start = 46, length = 12)

I don't think this is doing what you are saying here. It just means this call to the promise ignores the errors. The next usage of the promise could still have errors come out.